src/visitor.rs

```diff
@@ -0,0 +1,769 @@
+// Copyright (C) 2024 Parity Technologies (UK) Ltd. (admin@parity.io)
+// This file is a part of the scale-decode crate.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//         http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#![allow(clippy::type_complexity)]
+
+use crate::{
+    BitsOrderFormat, BitsStoreFormat, Field, FieldIter, Primitive, ResolvedTypeVisitor,
+    UnhandledKind, Variant, VariantIter,
+};
+use smallvec::SmallVec;
+
+/// A concrete iterator over struct or variant fields information.
+pub struct ConcreteFieldIter<'resolver, TypeId> {
+    fields: SmallVec<[Field<'resolver, TypeId>; 16]>,
+    idx: usize,
+}
+
+impl<'resolver, TypeId> Iterator for ConcreteFieldIter<'resolver, TypeId> {
+    type Item = Field<'resolver, TypeId>;
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.fields.get(self.idx) {
+            Some(field) => {
+                self.idx += 1;
+                Some(*field)
+            }
+            None => None,
+        }
+    }
+}
+
+impl<'resolver, TypeId> ExactSizeIterator for ConcreteFieldIter<'resolver, TypeId> {
+    fn len(&self) -> usize {
+        self.fields.len()
+    }
+}
+
+/// This is a concrete visitor which implements [`ResolvedTypeVisitor`]. It's instantiated by
+/// calling the **standalone** [`new`] function; see the docs there for more information.
+pub struct ConcreteResolvedTypeVisitor<
+    'resolver,
+    Context,
+    TypeId,
+    Output,
+    UnhandledFn,
+    NotFoundFn,
+    CompositeFn,
+    VariantFn,
+    SequenceFn,
+    ArrayFn,
+    TupleFn,
+    PrimitiveFn,
+    CompactFn,
+    BitSequenceFn,
+> {
```

Comment:
Question: Is it possible to box up this monstrosity as a `Box<dyn ResolvedTypeVisitor>` to store it or send it to different places? Or is `ResolvedTypeVisitor` not object safe? I imagine it might be hard to specify the type of a ConcreteResolvedTypeVisitor if this is ever needed.

Reply:
The concrete visitor object that we construct could be sent around and such (though I had no use cases for that; mostly you'd construct and use a visitor right next to a `types.resolve_type(type_id, visitor)` call.

I don't think any useful part of this impl would be object-safe given all of the generics! The inner functions could be boxed or put behind references to simplify (though I wouldn't want to box anything because it's meant to be zero cost as much as possible) but I didn't think it would be worth while given the cost!

Reply:
this should mainly accessed via `visitor::new(context, | .... |)` or am I misunderstanding anything?